x-pack/plugins/ingest_manager/server/routes/agent_config/handlers.ts

```diff
@@ -193,3 +222,35 @@ export const getFullAgentConfig: RequestHandler<TypeOf<
     });
   }
 };
+
+const buildSystemDatasource = async (
```

Comment:
It feels like this function belongs in one of the services as a method - maybe Datasources?
Let me know if its best for it to live elsewhere.

Reply:
I think this logic could be split up into two services.

1) A new service inside `x-pack/plugins/ingest_manager/common/services/package_to_config.ts` that builds a complete new datasource from package info:

```ts
export const packageToConfigDatasource = (
  packageInfo: PackageInfo,
  configId: string,
  outputId: string,
  datasourceName?: string
): NewDatasource => {
      return {
        name: datasourceName || `${packageInfo.name}-1`,
        package: {
          name: packageInfo.name,
          title: packageInfo.title,
          version: packageInfo.version,
        },
        enabled: true,
        config_id: configId,
        output_id: outputId,
        inputs: packageToConfigDatasourceInputs(packageInfo),
      };
}
```

2) A new datasourceService that retrieves the information needed to use that service inside `x-pack/plugins/ingest_manager/server/services/datasource.ts`:

(almost identical to what you currently have here, just removing `sys` and changing return line to use the generic service)

```ts
const createDatasourceFromPackage = async (
  soClient: SavedObjectsClientContract,
  pkgName: string,
): Promise<NewDatasource | undefined> => {
  const pkgInstall = await findInstalledPackageByName({
    savedObjectsClient: soClient,
    pkgName,
  });
  if (pkgInstall) {
    const [pkgInfo, defaultOutputId] = await Promise.all([
      getPackageInfo({
        savedObjectsClient: soClient,
        pkgkey: `${pkgInstall.name}-${pkgInstall.version}`,
      }),
      outputService.getDefaultOutputId(soClient),
    ]);
    if (pkgInfo) {
      return packageToConfigDatasource(pkgInfo, '', defaultOutputId);
    }
  }
};
```

What do you think? I can see the generic service inside `/common` being useful elsewhere too, hence my not putting everything inside `datasourceService`. It also makes it easier to write unit tests for.

Reply:
thanks for the feedback @jen-huang . This makes sense to me and I agree that that the common service to top-level `/common/services` will be helpful in a few places (maybe even for Endpoint as we will need to create datasources for Policy).  I have also noticed that we generate the package key manually in several places - maybe a service to handle that as well so that the generation of the value can be centralized?